RecoMuon/TrackerSeedGenerator/plugins/TSGForOIDNN.cc

```diff
@@ -0,0 +1,974 @@
+    } else {
+      // use classifier (default)
+      dnnModelPath_barrel_ = metadata.get<std::string>("barrel.dnnmodel_path");
+      edm::FileInPath dnnPath_barrel(dnnModelPath_barrel_);
+      graphDef_barrel_ = tensorflow::loadGraphDef(dnnPath_barrel.fullPath());
+      tf_session_barrel_ = tensorflow::createSession(graphDef_barrel_);
+
+      dnnModelPath_endcap_ = metadata.get<std::string>("endcap.dnnmodel_path");
+      edm::FileInPath dnnPath_endcap(dnnModelPath_endcap_);
+      graphDef_endcap_ = tensorflow::loadGraphDef(dnnPath_endcap.fullPath());
+      tf_session_endcap_ = tensorflow::createSession(graphDef_endcap_);
+    }
+  }
+  produces<std::vector<TrajectorySeed> >();
+}
+
+TSGForOIDNN::~TSGForOIDNN() {
+  if (getStrategyFromDNN_) {
+    if (useRegressor_) {
+      tensorflow::closeSession(tf_session_HB_);
+      tensorflow::closeSession(tf_session_HLIP_);
+      tensorflow::closeSession(tf_session_HLMuS_);
+      delete graphDef_HB_;
+      delete graphDef_HLIP_;
+      delete graphDef_HLMuS_;
+    } else {
+      tensorflow::closeSession(tf_session_barrel_);
+      tensorflow::closeSession(tf_session_endcap_);
+      delete graphDef_barrel_;
+      delete graphDef_endcap_;
+    }
+  }
+}
+
+//
+// Produce seeds
+//
+void TSGForOIDNN::produce(edm::StreamID sid, edm::Event& iEvent, edm::EventSetup const& iEventSetup) const {
+  // Initialize variables
+  unsigned int numSeedsMade = 0;
+  unsigned int layerCount = 0;
+  unsigned int hitlessSeedsMadeIP = 0;
+  unsigned int hitlessSeedsMadeMuS = 0;
+  unsigned int hitSeedsMade = 0;
+  unsigned int hitDoubletSeedsMade = 0;
+
+  // Surface used to make a TSOS at the PCA to the beamline
+  Plane::PlanePointer dummyPlane = Plane::build(Plane::PositionType(), Plane::RotationType());
+
+  // Read ESHandles
+  edm::ESHandle<Chi2MeasurementEstimatorBase> estimatorH = iEventSetup.getHandle(t_estimatorH_);
+  edm::ESHandle<MagneticField> magfieldH = iEventSetup.getHandle(t_magfieldH_);
+  edm::ESHandle<Propagator> propagatorAlongH = iEventSetup.getHandle(t_propagatorAlongH_);
+  edm::ESHandle<Propagator> propagatorOppositeH = iEventSetup.getHandle(t_propagatorOppositeH_);
+  edm::ESHandle<TrackerGeometry> tmpTkGeometryH = iEventSetup.getHandle(t_tmpTkGeometryH_);
+  edm::ESHandle<GlobalTrackingGeometry> geometryH = iEventSetup.getHandle(t_geometryH_);
+  edm::ESHandle<NavigationSchool> navSchool = iEventSetup.getHandle(t_navSchool_);
+
+  edm::Handle<MeasurementTrackerEvent> measurementTrackerH;
+  iEvent.getByToken(measurementTrackerTag_, measurementTrackerH);
+
+  // Read L2 track collection
+  edm::Handle<reco::TrackCollection> l2TrackCol;
+  iEvent.getByToken(src_, l2TrackCol);
+
+  // The product
+  std::unique_ptr<std::vector<TrajectorySeed> > result(new std::vector<TrajectorySeed>());
+
+  // Get vector of Detector layers
+  std::vector<BarrelDetLayer const*> const& tob = measurementTrackerH->geometricSearchTracker()->tobLayers();
```

Comment:
```suggestion
  auto const* gsTracker = measurementTrackerH->geometricSearchTracker();
  std::vector<BarrelDetLayer const*> const& tob = gsTracker->tobLayers();
```
and reuse in the 4 more cases below

Reply:
Done!